src/main/java/com/google/sps/servlets/HandlePhotoShoppingServlet.java

```diff
@@ -0,0 +1,184 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import com.google.appengine.api.blobstore.BlobInfo;
+import com.google.appengine.api.blobstore.BlobInfoFactory;
+import com.google.appengine.api.blobstore.BlobKey;
+import com.google.appengine.api.blobstore.BlobstoreService;
+import com.google.appengine.api.blobstore.BlobstoreServiceFactory;
+
+import com.google.gson.Gson;
+
+import com.google.sps.GoogleShoppingQuerier;
+import com.google.sps.ShoppingQuerierConnectionException;
+import com.google.sps.data.Product;
+import com.google.sps.data.ShoppingQueryInput;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import static javax.servlet.http.HttpServletResponse.SC_INTERNAL_SERVER_ERROR;
+
+/**
+ * When the user submits the form for image uploading, Blobstore processes the file upload and forwards
+ * the request to this servlet, which returns the product shopping results for the respective photo,
+ * in JSON format, along with the shopping query used to search.
+ */
+@WebServlet("handle-photo-shopping")
+public class HandlePhotoShoppingServlet extends HttpServlet {
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Get the BlobKey that points to the image uploaded by the user.
+    BlobKey uploadedImageBlobKey = getBlobKey(request, "photo");
+
+    // Send an error if the user did not upload a file.
+    if (uploadedImageBlobKey == null) {
+      response.sendError(HttpServletResponse.SC_BAD_REQUEST, "Client must upload an image file.");
+      return;
+    }
+
+    // Get the photo category (i.e. product, shopping-list or barcode) entered by the user.
+    if (request.getParameter("photo-category").isEmpty()) {
+      response.sendError(
+          HttpServletResponse.SC_BAD_REQUEST, "Client must select a photo category when submitting the form.");
```

Comment:
This message could also be as simple as "missing a photo category".

Try to keep the message simple and direct so that it's easier/quicker for clients to understand.